File: src/main/cpp/RobotContainer.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "RobotContainer.h"

#include <frc2/command/button/Trigger.h>

#include "commands/Autos.h"
#include "commands/ExampleCommand.h"
#include "commands/command_RunMotor.h"
#include "commands/command_ManualMotor.h"


RobotContainer::RobotContainer() {
  // Initialize all of your commands and subsystems here
  // m_motor.SetDefaultCommand(command_RunMotor(&m_motor, [this]{return m_drivercontroller;})); //FIX THIS WITH XBOX mappings
  // m_motor.SetDefaultCommand(command_RunMotor(&m_motor, [=] {return WristPositions::HighConePos;}, [=] {return false;}, 
  //                                           [=] {return WristConstants::WristThreshold;}).ToPtr());
  // // Configure the button bindings
  // m_motor.ResetWrist();
  m_motor.SetDefaultCommand(command_ManualMotor(&m_motor,
                          [this]{return -m_driverController.GetRawAxis(OperatorConstants::xboxLYAxis);}//replace "this" with = if problem
                       ).ToPtr());
  // frc::XboxController::Axis::
  ConfigureBindings();
}
 void RobotContainer::ConfigureBindings() {
  // Configure your trigger bindings here
  // Schedule `ExampleCommand` when `exampleCondition` changes to `true`
   frc2::JoystickButton(&m_driverController, 
                       frc::XboxController::Button::kY)
                       .OnTrue(command_RunMotor(&m_motor, 
                                            [=] {return WristPositions::HighConePos;}, 
                                            [=] {return false;}, 
                                            [=] {return WristConstants::WristThreshold;},
                                            [=] {return 0;}
                       ).ToPtr());
  
  frc2::JoystickButton(&m_driverController, 
                       frc::XboxController::Button::kX)
                       .OnTrue(command_RunMotor(&m_motor, 
                                            [=] {return WristPositions::HighCubePos;}, 
                                            [=] {return false;}, 
                                            [=] {return WristConstants::WristThreshold;},
                                            [=] {return 0;}
                       ).ToPtr());

  frc2::JoystickButton(&m_driverController, 
                       frc::XboxController::Button::kB)
                       .OnTrue(command_RunMotor(&m_motor, 
                                            [=] {return WristPositions::MidConePos;}, 
                                            [=] {return false;}, 
                                            [=] {return WristConstants::WristThreshold;},
                                            [=] {return 0;}
                       ).ToPtr());

  frc2::JoystickButton(&m_driverController, 
                       frc::XboxController::Button::kA)
                       .OnTrue(command_RunMotor(&m_motor, 
                                            [=] {return WristPositions::MidCubePos;}, 
                                            [=] {return false;}, 
                                            [=] {return WristConstants::WristThreshold;},
                                            [=] {return 0;}
                       ).ToPtr());
                      //add another button mapping for substation
  frc2::JoystickButton(&m_driverController,
                        frc::XboxController::Button::kLeftBumper)
                        .OnTrue(command_RunMotor(&m_motor, 
                                            [=] {return WristPositions::SubstationPos;}, 
                                            [=] {return false;}, 
                                            [=] {return WristConstants::WristThreshold;},
                                            [=] {return 0;}
                       ).ToPtr());

  frc2::JoystickButton(&m_driverController,
                        frc::XboxController::Button::kRightBumper)
                        .OnTrue(command_RunMotor(&m_motor, 
                                            [=] {return 0;}, //stow pos
                                            [=] {return false;}, 
                                            [=] {return WristConstants::WristThreshold;},
                                            [=] {return 0;}
                       ).ToPtr());
    frc2::JoystickButton(&m_driverController,
                        frc::XboxController::Button::kLeftStick)
                        .OnTrue(command_RunMotor(&m_motor, 
                                            [=] {return WristPositions::GroundIntake;}, 
                                            [=] {return false;}, 
                                            [=] {return WristConstants::WristThreshold;},
                                            [=] {return 0;}
                       ).ToPtr());
  
  // frc2::JoystickButton(&m_driverController, use this for reset writst 
  //                       frc::XboxController::Button::kRightBumper)
  //                       .OnTrue(command_RunMotor(&m_motor, 
  //                                           [=] {return WristPositions::HighConePos;}, 
  //                                           [=] {return true;}, 
  //                                           [=] {return WristConstants::wristThreshold;}
  //                      ).ToPtr());
  // Schedule `ExampleMethodCommand` when the Xbox controller's B button is
  // pressed, cancelling on release.
  // m_driverController.B().WhileTrue(m_subsystem.ExampleMethodCommand());
}

frc2::CommandPtr RobotContainer::GetAutonomousCommand() {
  // An example command will be run in autonomous
  return autos::ExampleAuto(&m_subsystem);
}



//control guide Y is high cone, X is high cube, B is mid cone, A is mid cube, left bumper is double sub,
//right bumper is stow
//left stick is ground intake

File: src/main/java/frc/robot/RobotContainer.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import edu.wpi.first.wpilibj.GenericHID;
import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.XboxController;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.InstantCommand;
import edu.wpi.first.wpilibj2.command.button.JoystickButton;
// import frc.robot.autos.*;
import frc.robot.commands.*;
import frc.robot.subsystems.*;
import frc.robot.Constants.ControllerConstants;

/**
 * This class is where the bulk of the robot should be declared. Since Command-based is a
 * "declarative" paradigm, very little robot logic should actually be handled in the {@link Robot}
 * periodic methods (other than the scheduler calls). Instead, the structure of the robot (including
 * subsystems, commands, and button mappings) should be declared here.
 */
public class RobotContainer {
  /* Controllers */
  private final Joystick driveRad = new Joystick(ControllerConstants.driveRadID);
  private final Joystick haPerator = new Joystick(ControllerConstants.xboxHaperatorID);

  /* Drive Controls */
  private final int translationAxis = ControllerConstants.xboxLYAxis;
  private final int strafeAxis = ControllerConstants.xboxLXAxis;
  private final int rotationAxis = ControllerConstants.xboxRXAxis;

  private final int zeroGyroButton = ControllerConstants.xboxY;
  private final int robotCentricButton = ControllerConstants.xboxLB;
  private final int zeroAngleButton = ControllerConstants.xboxA;

  /* Driver Buttons */
  private final JoystickButton zeroGyro =
      new JoystickButton(driveRad, zeroGyroButton);
  private final JoystickButton robotCentric =
      new JoystickButton(driveRad, robotCentricButton);
  private final JoystickButton zeroAngle = 
      new JoystickButton(driveRad, zeroAngleButton);

  /* Operator Controls */
  private final int wristAxis = ControllerConstants.xboxLYAxis;
  private final int elevatorAxis = ControllerConstants.xboxRYAxis;

  private final int highConeButton = ControllerConstants.xboxY;
  private final int midConeButton = ControllerConstants.xboxB;
  private final int highCubeButton = ControllerConstants.xboxX;
  private final int midCubeButton = ControllerConstants.xboxA;

  private final int toggleLEDButton = ControllerConstants.xboxLeftJoyPress;
  private final int doubleSubButton = ControllerConstants.xboxLB;
  private final int groundIntakeButton = ControllerConstants.xboxView; // a.k.a Back Button
  private final int stowButton = ControllerConstants.xboxMenu; // a.k.a Start Button
  
  /* Operator Buttons */
  private final JoystickButton highCone =
      new JoystickButton(haPerator, highConeButton
      );
  private final JoystickButton midCone =
      new JoystickButton(haPerator, midConeButton
      );
  private final JoystickButton highCube =
      new JoystickButton(haPerator, highCubeButton
      );
  private final JoystickButton midCube =
      new JoystickButton(haPerator, midCubeButton
      );
  private final JoystickButton toggleLED =
      new JoystickButton(haPerator, toggleLEDButton
      );
  private final JoystickButton doubleSub =
      new JoystickButton(haPerator, doubleSubButton
      );
  private final JoystickButton groundIntake =
      new JoystickButton(haPerator, groundIntakeButton
      );
  private final JoystickButton stow =
      new JoystickButton(haPerator, stowButton
      );

  /* Subsystems */
  private final subsystem_DriveTrain s_Swerve = new subsystem_DriveTrain();
  private final subsystem_Elevator s_Elevator = new subsystem_Elevator();
  private final subsystem_Wrist s_Wrist = new subsystem_Wrist();
  private final subsystem_Intake s_Intake = new subsystem_Intake();
  private final subsystem_LEDs s_LED = new subsystem_LEDs();

  /** The container for the robot. Contains subsystems, OI devices, and commands. */
  public RobotContainer() {
    s_Swerve.setDefaultCommand(
        new command_DriveTeleop(
            s_Swerve,
            () -> {return -driveRad.getRawAxis(translationAxis);},
            () -> {return -driveRad.getRawAxis(strafeAxis);},
            () -> {return -driveRad.getRawAxis(rotationAxis);},
            () -> {return robotCentric.getAsBoolean();}));

    // s_Elevator.setDefaultCommand(
    //     new command_MoveElevatorManually(s_Elevator, 
    //     () -> {return -haPerator.getRawAxis(elevatorAxis);})
    // );

    // s_Wrist.setDefaultCommand(
    //   new command_MoveWristManually(s_Wrist, 
    //   () -> {return -haPerator.getRawAxis(wristAxis);})
    // );
    
    // Configure the button bindings
    configureButtonBindings();
  }

  /**
   * Use this method to define your button->command mappings. Buttons can be created by
   * instantiating a {@link GenericHID} or one of its subclasses ({@link
   * edu.wpi.first.wpilibj.Joystick} or {@link XboxController}), and then passing it to a {@link
   * edu.wpi.first.wpilibj2.command.button.JoystickButton}.
   */
  private void configureButtonBindings() {
    /* Driver Buttons */
    // zeroGyro.whenPressed(new InstantCommand(() -> s_Swerve.zeroGyro()));
    zeroGyro.onTrue(new InstantCommand(() -> s_Swerve.zeroGyro()));

    highCone.onTrue(cGroup_Elevator.scoreHighCone(s_Elevator, s_Wrist, s_Intake, s_LED));

    midCone.onTrue(cGroup_Elevator.scoreMidCone(s_Elevator, s_Wrist, s_Intake, s_LED));

    highCube.onTrue(cGroup_Elevator.scoreHighCube(s_Elevator, s_Wrist, s_Intake, s_LED));
    
    midCube.onTrue(cGroup_Elevator.scoreMidCube(s_Elevator, s_Wrist, s_Intake, s_LED));
    
    toggleLED.onTrue(s_LED.cycleLEDCommand());

    doubleSub.onTrue(cGroup_Elevator.doubleSubstationIntake(s_Elevator, s_Wrist, s_Intake, s_LED));

    groundIntake.onTrue(cGroup_Elevator.groundIntake(s_Elevator, s_Wrist, s_Intake, s_LED));

    stow.onTrue(cGroup_Elevator.toStow(s_Elevator, s_Wrist));

    zeroAngle.onTrue(s_Swerve.zeroModuleAngles());
    }

  /**
   * Use this to pass the autonomous command to the main {@link Robot} class.
   *
   * @return the command to run in autonomous
   */
  public Command getAutonomousCommand() {
    // An ExampleCommand will run in autonomous
    return new Autos(s_Swerve);
  }
}
